wip/beam-spec.md

```diff
@@ -0,0 +1,437 @@
+# The Basic Embedded Abstract Machine (BEAM)
+
+## Summary
+
+This document contains the specification of an *abstract machine*. The goal of
+this specification is facilitating discussions about the soundness of `unsafe`
+concurrent code that runs on microcontrollers.
+
+The BEAM is meant to be representative of the microcontrollers one can run Rust
+code on, as such it has concurrency capabilities commonly found in this kind of
+hardware but exposes these capabilities in a way that's not tied to any
+particular architecture.
+
+## Program execution
+
+- The BEAM has a single core / processing-unit and can only run a single process
+  on this core.
+
+- When the BEAM executes a program it uses the function called `main` as the
+  entry point. Returning from this function is UNSPECIFIED behavior.
+
+- The signature of `main` must be `[unsafe] fn() [-> !]`. That is the program
+  takes no arguments.
+
+## Interrupt handling
+
+(Interrupt handling is a preemption mechanism that in its most basic form works
+as follows: when an interrupt *signal* arrives the processor suspends the
+execution of the current function and jumps to a special subroutine known as the
+interrupt *handler*. When this subroutine returns the processors resumes the
+execution of the function that was preempted by the interrupt handler.)
+
+- BEAM's interrupt handling capabilities can be summarized as: prioritization of
+  interrupts, individual interrupt masking and global interrupt masking.
+
+- The BEAM has 8 *interrupts* whose *handlers* are named: `INTERRUPT0`,
+  `INTERRUPT1`, all the way to `INTERRUPT7`.
+
+- The signature of all interrupt handlers must be `[unsafe] fn()`. That is these
+  handlers don't take arguments nor can they return values.
```

Comment:
To me this document feels like a step towards specifying what are interrupt handlers allowed / not allowed to do. 

I think that is worth specifying and we might be able to do that without having to speak about "masking" and many other concepts touched in this document. 

From that point-of-view, this single point is one of the most important points of the document to me. 

This point mentions that their type must be `fn() -> (), yet if they are not defined, they are implemented as `fn interrupt() { loop {} }`, so I wonder, can their type also be `fn() -> !` ? 

Also, if they return, they either have no effects (e.g. `fn interrupt() {}`) or they have a side-effect. Which side-effects are they allowed to have ? For example, suppose we have:

```rust
static PTR: UnsafeCell<*mut i32>;
unsafe fn interrupt() { PTR.get().write_volatile(42); }
fn foo() {
    let mut x = 0;
    PTR.set(&mut x);
    // preempted to `interrupt` here
    dbg!(x);  // Does this print `0` or `42` ?
}
```

`foo` does not contain any `unsafe { }` block, yet its execution can be preemtpted to an `unsafe fn` call, which can modify `foo` stack. Can `dbg!(x)` be optimized by the compiler to `dbg!(0)` ?

Reply:
In a sense, `interrupt`s always run in a different thread of execution, so even thought the application is single threaded, they should move into them any types that are not `Send`, should not read / write through references to types that are not `Sync`, etc. 

Ideally, we would have an `#[interrupt]` annotation in the language that would enforce these invariants at the type level and produce proper errors.

Reply:
> In a sense, interrupts always run in a different thread of execution,

I agree that this is the best way to think about them.
In user space, the same happens with signal handlers.